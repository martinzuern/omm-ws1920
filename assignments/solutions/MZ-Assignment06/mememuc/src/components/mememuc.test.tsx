import React from 'react';
import ReactDOM from 'react-dom';
import { render, fireEvent, screen, wait } from '@testing-library/react';

import OmmMememuc from './mememuc';

it('renders without crashing', async () => {
  const div = document.createElement('div');
  ReactDOM.render(<OmmMememuc />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders meme correctly', async () => {
  const { container } = render(<OmmMememuc />);

  // Wait for memes to be loaded from API
  await wait(() => screen.getByAltText('raptor meme'));

  // get the meme
  fireEvent.click(screen.getByAltText('doge meme'));

  // fill out the captions
  fireEvent.change(screen.getByPlaceholderText('Upper Text'), { target: { value: 'ate too many chickem nuggers' } });
  fireEvent.change(screen.getByPlaceholderText('Lower Text'), { target: { value: 'need to sleep plz' } });

  expect(container.querySelector('.results')).toMatchSnapshot();
});

describe('Test meme handling', () => {
  // Note:
  // Bad practice to test one component through another, but for simplity I want to keep everything in this file.

  let containerDOM: HTMLElement;
  beforeEach(async () => {
    const { container } = render(<OmmMememuc />);
    containerDOM = container;

    // Wait for memes to be loaded from API
    await wait(() => screen.getByAltText('raptor meme'));

    const saveMeme = (meme: string, text1: string, text2: string) => {
      fireEvent.click(screen.getByAltText(meme));
      fireEvent.change(screen.getByPlaceholderText('Upper Text'), { target: { value: text1 } });
      fireEvent.change(screen.getByPlaceholderText('Lower Text'), { target: { value: text2 } });
      fireEvent.click(screen.getByText('Save to favorites'));
    };

    saveMeme('doge meme', 'ate too many chickem nuggers', 'need to sleep plz');
    saveMeme('boromir meme', 'One does not simply', 'skip the assignments');
    saveMeme('raptor meme', "Isn't the raptor claw", 'on the foot?');
  });

  it('saves meme to favorites', async () => {
    expect(containerDOM.querySelector('.meme-history')).toMatchSnapshot();
  });

  it('allows meme to be restored from favorites', async () => {
    const resultsBefore = containerDOM.querySelector('.results');
    expect(resultsBefore).toMatchSnapshot();

    const firstMeme = containerDOM.querySelector('.a-meme-history-item img');
    expect(firstMeme).toBeTruthy();
    fireEvent.click(firstMeme!);

    const resultsAfter = containerDOM.querySelector('.results');
    expect(resultsAfter).toMatchSnapshot();
  });

  it('allows meme to be removed from favorites', async () => {
    expect(containerDOM.querySelectorAll('.a-meme-history-item').length).toBe(3);

    // Get second meme
    const secondMeme = containerDOM.querySelector('.meme-history .a-meme-history-item:nth-child(2) .mat-icon');
    expect(secondMeme).toBeTruthy();
    fireEvent.click(secondMeme!);

    expect(containerDOM.querySelectorAll('.a-meme-history-item').length).toBe(2);
    expect(containerDOM.querySelector('.meme-history')).toMatchSnapshot();
  });
});
