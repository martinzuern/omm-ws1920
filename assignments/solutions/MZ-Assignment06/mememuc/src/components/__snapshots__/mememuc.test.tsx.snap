// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test meme handling allows meme to be removed from favorites 1`] = `
<div
  class="meme-history"
>
  <div
    class="a-meme-history-item"
  >
    <img
      alt="history"
      src="http://212.227.193.27:3000/memes/doge?text=ate+too+many+chickem+nuggers&x=0&y=0&text2=need+to+sleep+plz&x2=0&y2=0"
    />
    <div
      class="mat-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
  <div
    class="a-meme-history-item"
  >
    <img
      alt="history"
      src="http://212.227.193.27:3000/memes/raptor?text=Isn%27t+the+raptor+claw&x=0&y=0&text2=on+the+foot%3F&x2=0&y2=0"
    />
    <div
      class="mat-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Test meme handling allows meme to be restored from favorites 1`] = `
<div
  class="results"
>
  <img
    alt="selected"
    src="http://212.227.193.27:3000/memes/raptor?text=Isn%27t+the+raptor+claw&x=0&y=0&text2=on+the+foot%3F&x2=0&y2=0"
  />
</div>
`;

exports[`Test meme handling allows meme to be restored from favorites 2`] = `
<div
  class="results"
>
  <img
    alt="selected"
    src="http://212.227.193.27:3000/memes/doge?text=ate+too+many+chickem+nuggers&x=0&y=0&text2=need+to+sleep+plz&x2=0&y2=0"
  />
</div>
`;

exports[`Test meme handling saves meme to favorites 1`] = `
<div
  class="meme-history"
>
  <div
    class="a-meme-history-item"
  >
    <img
      alt="history"
      src="http://212.227.193.27:3000/memes/doge?text=ate+too+many+chickem+nuggers&x=0&y=0&text2=need+to+sleep+plz&x2=0&y2=0"
    />
    <div
      class="mat-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
  <div
    class="a-meme-history-item"
  >
    <img
      alt="history"
      src="http://212.227.193.27:3000/memes/boromir?text=One+does+not+simply&x=0&y=0&text2=skip+the+assignments&x2=0&y2=0"
    />
    <div
      class="mat-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
  <div
    class="a-meme-history-item"
  >
    <img
      alt="history"
      src="http://212.227.193.27:3000/memes/raptor?text=Isn%27t+the+raptor+claw&x=0&y=0&text2=on+the+foot%3F&x2=0&y2=0"
    />
    <div
      class="mat-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders meme correctly 1`] = `
<div
  class="results"
>
  <img
    alt="selected"
    src="http://212.227.193.27:3000/memes/doge?text=ate+too+many+chickem+nuggers&x=0&y=0&text2=need+to+sleep+plz&x2=0&y2=0"
  />
</div>
`;
